nomad/structs/csi_volumes.go

```diff
@@ -0,0 +1,248 @@
+package structs
+
+import (
+	"fmt"
+	"strings"
+	"time"
+)
+
+const (
+	VolumeTypeCSI = "csi"
+)
+
+type CSIVolume struct {
+	ID         string
+	Driver     string
+	Namespace  string
+	Topology   *CSITopology
+	MaxReaders int
+	MaxWriters int
+
```

Comment:
We also need to store the desired Access Mode (SingleNodeReadOnly/SingleNodeReadWrite/MultiNodeSingleReader etc) and the desired Attachment Mode here

Reply:
Doesn't the desired attachment mode go with the job task? I think we can mount the same volume writable for one taskgroup and readonly for many.

My plan for allowed access modes here is that a readonly volume is expressed with `MaxWriters: 0`. Future support for multiple writers can just keep using this bit of api.

Reply:
I think we want to explicitly model the access modes at registration time tbh, otherwise we need to infer the correct one at attachment time (which is ~fine, but in some cases could have massive performance differences if we infer the wrong one)

Reply:
I added the modes. I moved your definitions from `client/structs` copying the direction of other dependencies, not 100% sure that's correct.